x-pack/plugin/ml/src/main/java/org/elasticsearch/xpack/ml/job/process/normalizer/ShortCircuitingRenormalizer.java

```diff
@@ -51,35 +49,32 @@ public void renormalize(Quantiles quantiles) {
             return;
         }
 
-        // This will throw NPE if quantiles is null, so do it first
-        QuantilesWithLatch quantilesWithLatch = new QuantilesWithLatch(quantiles, new CountDownLatch(1));
         // Needed to ensure work is not added while the tryFinishWork() method is running
-        synchronized (quantilesDeque) {
-            // Must add to latchDeque before quantilesDeque
-            latchDeque.addLast(quantilesWithLatch.getLatch());
-            quantilesDeque.addLast(quantilesWithLatch);
-            executorService.submit(() -> doRenormalizations());
+        synchronized (this) {
+            latestQuantilesHolder = (latestQuantilesHolder == null)
+                ? new AugmentedQuantiles(quantiles, null, new CountDownLatch(1))
+                : new AugmentedQuantiles(quantiles, latestQuantilesHolder.getEvictedTimestamp(), latestQuantilesHolder.getLatch());
         }
+        executorService.submit(this::doRenormalizations);
     }
 
     @Override
-    public void waitUntilIdle() {
-        try {
-            // We cannot tolerate more than one thread running this loop at any time,
-            // but need a different lock to the other synchronized parts of the code
-            synchronized (latchDeque) {
-                for (CountDownLatch latchToAwait = latchDeque.pollFirst(); latchToAwait != null; latchToAwait = latchDeque.pollFirst()) {
-                    latchToAwait.await();
-                }
+    public void waitUntilIdle() throws InterruptedException {
+        CountDownLatch latch;
+        do {
+            synchronized (this) {
+                latch = (latestQuantilesHolder != null) ? latestQuantilesHolder.getLatch() : null;
             }
-        } catch (InterruptedException e) {
```

Comment:
> It's like we need a mechanism to ask if we could acquire the semaphore without actually acquiring it.

OK, there is an `availablePermits()` method that does this, but the problem is that then we'd be doing a busy wait instead of waiting for a wakeup from the OS.

We need something other than a `Semaphore`.  Maybe a `Phaser` or `CyclicBarrier`.  I will have another go at it.